stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py

```diff
@@ -0,0 +1,280 @@
+            'product_uom': self.product_uom.id,
+            'restrict_lot_id': self.lot_id.id
+            if self.lot_id else False,
+            }
+
+
+class StockPickingPackagePreparation(models.Model):
+
+    _inherit = 'stock.picking.package.preparation'
+
+    FIELDS_STATES = {'done': [('readonly', True)],
+                     'in_pack': [('readonly', True)],
+                     'cancel': [('readonly', True)]}
+
+    picking_type_id = fields.Many2one(related='picking_ids.picking_type_id')
+    line_ids = fields.One2many(
+        'stock.picking.package.preparation.line',
+        'package_preparation_id',
+        string='Details',
+        copy=False,
+        states=FIELDS_STATES)
+
+    @api.multi
+    def _update_line_ids(self):
+        # Update package lines according to pickings.
+        # Lines with move_id will be updated
+        # Lines without move_id not be touched
+        line_model = self.env['stock.picking.package.preparation.line']
+        for pack in self:
+            for line in pack.line_ids:
+                if line.move_id:
+                    line.unlink()
+            package_preparation_lines = (
+                line_model._prepare_lines_from_pickings(
+                    [p.id for p in pack.picking_ids]))
+            if package_preparation_lines:
+                for line_vals in package_preparation_lines:
+                    line_vals['package_preparation_id'] = pack.id
+                    line_model.create(line_vals)
+
+    @api.model
+    @api.returns('self', lambda value: value.id)
+    def create(self, values):
+        values = self._update_line_vals(values)
+        pack = super(StockPickingPackagePreparation, self).create(values)
+        return pack
+
+    def _update_line_vals(self, values):
+        # ----- Create a PackagePreparationLine for every stock move
+        #       in the pickings added to PackagePreparation
+        if values.get('picking_ids', False):
+            picking_ids = []
+            for pick_tuple in values['picking_ids']:
+                if pick_tuple[0] == 6:
+                    picking_ids.extend(pick_tuple[2])
+                elif pick_tuple[0] == 4:
+                    picking_ids.append(pick_tuple[1])
+            package_preparation_lines = self.env[
+                'stock.picking.package.preparation.line'
+                ]._prepare_lines_from_pickings(picking_ids)
+            if package_preparation_lines:
+                values.update({
+                    'line_ids': [(0, 0, v) for v in package_preparation_lines]
+                })
+        return values
+
+    @api.multi
+    def write(self, values):
+        # ----- Delete package preparation line if the relative picking is
+        #       delete from package preparation
+        if values.get('picking_ids', False):
+            package_preparation_line_model = self.env[
+                'stock.picking.package.preparation.line']
+            # ----- Collect new pickings to read the change
```

Comment:
Understood - I won't block over this, but I recommend removing. They add nothing to the code other than more maintenance. 

I am blocking on the Doc block comments though. Doc blocks are described in [PEP8](https://www.python.org/dev/peps/pep-0008/#documentation-strings) & [PEP237](https://www.python.org/dev/peps/pep-0257/)